src/Storages/ExternalStream/Kafka/Kafka.h

```diff
@@ -36,37 +60,53 @@ class Kafka final : public StorageExternalStreamImpl
     SinkToStoragePtr write(const ASTPtr & query, const StorageMetadataPtr & metadata_snapshot, ContextPtr context) override;
 
     bool produceOneMessagePerRow() const { return settings->one_message_per_row; }
+    Int32 topicRefreshIntervalMs() const { return topic_refresh_interval_ms; }
     const String & brokers() const { return settings->brokers.value; }
     const String & dataFormat() const override { return data_format; }
     const String & topic() const { return settings->topic.value; }
-    const klog::KConfParams & properties() const { return kafka_properties; }
-    const klog::KafkaWALAuth & auth() const noexcept { return *auth_info; }
     const ASTPtr & shardingExprAst() const { assert(!engine_args.empty()); return engine_args[0]; }
     bool hasCustomShardingExpr() const;
-    klog::KafkaWALSimpleConsumerPtr getConsumer(int32_t fetch_wait_max_ms = 200) const;
+
+    RdKafka::Producer & getProducer();
+    RdKafka::Topic & getProducerTopic();
+
+    RdKafka::ConsumerPool::Entry getConsumer() const
+    {
+        assert(consumer_pool);
+        return consumer_pool->get(/*max_wait_ms=*/1000);
+    }
+
+    String getLoggerName() const { return storage_id.getDatabaseName() == "default" ? storage_id.getTableName() : storage_id.getFullNameNotQuoted(); }
 
 private:
     void calculateDataFormat(const IStorage * storage);
     void cacheVirtualColumnNamesAndTypes();
     std::vector<Int64> getOffsets(const SeekToInfoPtr & seek_to_info, const std::vector<int32_t> & shards_to_query) const;
     void validateMessageKey(const String & message_key, IStorage * storage, const ContextPtr & context);
-    void validate(const std::vector<int32_t> & shards_to_query = {});
+    void validate();
     static std::vector<int32_t> parseShards(const std::string & shards_setting);
 
     StorageID storage_id;
     ASTs engine_args;
-    klog::KConfParams kafka_properties;
     String data_format;
-    const std::unique_ptr<klog::KafkaWALAuth> auth_info;
     ExternalStreamCounterPtr external_stream_counter;
 
     NamesAndTypesList virtual_column_names_and_types;
 
-    std::mutex shards_mutex;
-    int32_t shards = 0;
-
     ASTPtr message_key_ast;
+    Int32 topic_refresh_interval_ms = 0;
+    std::vector<Int32> shards_from_settings;
+
+    ConfPtr conf;
+    /// The Producer instance and Topic instance can be used by multiple sinks at the same time, thus we only need one of each.
+    std::mutex producer_mutex;
+    std::unique_ptr<RdKafka::Producer> producer;
+    std::unique_ptr<RdKafka::Topic> producer_topic;
```

Comment:
`topic`

Reply:
Same as the previous one.

Reply:
In my understanding, each producer and consumer has its own `Topic`, we can move the ownership of `Topic` to `Producer` and `Consumer`

Reply:
Resolving this comment since it's talking the same thing as the previous ones.